### ISSUE TITLE
Send does not return an error after stream is closed

### ISSUE THREAD

**Issue by bradylove** [grpc/grpc-go#1790]
### What version of gRPC are you using?

1.6.1 (Also tested on 1.9.1)

### What version of Go are you using (`go version`)?

1.9.2

### What operating system (Linux, Windows, …) and version?

Linux (Ubuntu 16.04)

### What did you do?

Opened a stream to a server that is not registered. I have created a [modified version of the helloworld example](https://gist.github.com/bradylove/87f621a8bbf1fef80fff1dd051d53087) to show this. 

Our end goal is to be able to detect whether a new API is available on a service, if it is use that API, otherwise fall back to our old deprecated API. We want to be able to do this for an HA deploy.

### What did you expect to see?

An error similar to what you would receive if you were to try to make an RPC call that does not exist on a registered server which is an error with code `12` (`Unimplemented`). 

### What did you see instead?

No errors and everything is successful. 

Is there a recommended way to detect if a server is registered or not without using reflection?

**Comment by dfawley** [comment]
If you modify your client to do an `r.CloseAndRecv()`, you will see the `Unimplemented` error.  What happens is this:

1. Client dials server
1. Server accepts connection
1. Client initiates RPC (sends headers)
1. Client application calls `Send()` to queue up a message; `Send` returns after flow control checks ensure it can be sent
  -- This is where your client terminates --
1. Serve receives the header referring to an unregistered RPC, and ends the stream with trailers that include status=`Unimplemented`
1. Client receives trailers

However, if those trailers are never read via a `Recv()` (or in this case, since it's client-streaming, `CloseAndRecv()`), the application will never see them.

I think there may also be an API problem here.  If the server terminates the stream, subsequent Send calls should arguably fail.  This would introduce a race between 6 and 4, above, and it would be a behavior change, so I'm not sure if this is something we should/can change.  I'll have to think more about it.

May I ask why you don't want to use reflection?

**Comment by bradylove** [comment]
@dfawley thank you for your feedback.

The reason I want to avoid reflection is because it is not currently enabled on the servers (now wishing we had enabled it already).

With your feedback I have this working with the following:

1. Open stream to new API
1. Call `CloseAndRecv()`
1. Check status code.
  a. If status code == 12, open stream deprecated API
  b. If status code == 0, open new stream to new API

Do you see any issues with this approach?

**Comment by dfawley** [comment]
That seems reasonable to me.  Obviously the server application could theoretically return an `Unimplemented` error from a registered handler, but if you know it doesn't, then that should work fine.

However, I think this is a bug.  We should return an error (`io.EOF`) from `Send` when the stream has been terminated by the server.  `Send` actually _will_ return an error today, but only after the flow control is exhausted.  We will fix this issue, but I think your idea above is fine as a workaround until that happens.

Note that because the network is asynchronous, there will still be an inherent race between steps 4 and 6, so you will not necessarily get an error upon the _first_ write, just the first write after the `Unimplemented` is received.

**Comment by dfawley** [comment]
Actually, this may be even worse than I previously thought.  If we are deducting these `Send`s from the transport's flow control, but then not writing the data on the stream and not adding our quota back to the flow control, then we will eventually run out of flow control on the connection.

**Comment by dfawley** [comment]
> If we are deducting these Sends from the transport's flow control, but then not writing the data on the stream

It looks like this is not the case; we are unconditionally writing the data to the stream even after it was terminated, so the server will send flow control window updates, and we will not run out of send quota.

**Comment by dfawley** [comment]
Actually, this bug is a duplicate of #1755, so I'll close this now.

The tl;dr here is that after #1792 is submitted, `Send()` itself will eventually give you an error (`io.EOF`), at which point `Recv()` will let you know it was `Unimplemented`.  So you shouldn't have to do a probing no-payload use of the RPC as suggested [above](https://github.com/grpc/grpc-go/issues/1790#issuecomment-356686985).
